In [ ]:
Pandas:
Panel Datas:

> Individual data, Institutional data, Domain data

In [6]:
# tools which contain structure data
# excel(semi-structured)
# csv (ustructured)
# database MYSQL, SQLIte, SQLSERVER, DB2, etc...(pure structured data)
# HTML(ustructured)
# XML
# WML
# etc...

In [7]:
# EDA:(Exploratory data analysis):

# Data filtering, manupulation,remove outliers,etc..

In [8]:
# Installation of pandas

In [1]:
import pandas as pd

In [2]:
# csv is a plain text file.--> get help to launch from excel sheet

In [3]:
a=pd.read_csv("test.txt")  # dataframe

In [4]:
a

SN     Name  Age
0   1  Krishna   27
1   2   Zulqar   21
2   3   Manish   27

In [5]:
print(type(a))

<class 'pandas.core.frame.DataFrame'>


In [6]:
df=pd.read_excel("wheather.xlsx")

In [9]:
df

Date  temperature  windspeed  event
0 2022-01-01         32.0        6.0   Rain
1 2022-01-02          NaN        9.0  Sunny
2 2022-01-03         28.0        NaN   Snow
3 2022-01-04          NaN        7.0    NaN
4 2022-01-05         32.0        NaN   Rain
5 2022-01-06          NaN        NaN  Sunny
6 2022-01-07          NaN        NaN    NaN
7 2022-01-08         34.0        8.0  Cloud
8 2022-01-09         40.0       12.0  Sunny
9 2022-01-10         30.0       10.0  Sunny

In [8]:
df.dropna()

Date  temperature  windspeed  event
0 2022-01-01         32.0        6.0   Rain
7 2022-01-08         34.0        8.0  Cloud
8 2022-01-09         40.0       12.0  Sunny
9 2022-01-10         30.0       10.0  Sunny

In [10]:
df

Date  temperature  windspeed  event
0 2022-01-01         32.0        6.0   Rain
1 2022-01-02          NaN        9.0  Sunny
2 2022-01-03         28.0        NaN   Snow
3 2022-01-04          NaN        7.0    NaN
4 2022-01-05         32.0        NaN   Rain
5 2022-01-06          NaN        NaN  Sunny
6 2022-01-07          NaN        NaN    NaN
7 2022-01-08         34.0        8.0  Cloud
8 2022-01-09         40.0       12.0  Sunny
9 2022-01-10         30.0       10.0  Sunny

In [13]:
df.temperature

0    32.0
1     NaN
2    28.0
3     NaN
4    32.0
5     NaN
6     NaN
7    34.0
8    40.0
9    30.0
Name: temperature, dtype: float64

In [14]:
df["temperature"]

0    32.0
1     NaN
2    28.0
3     NaN
4    32.0
5     NaN
6     NaN
7    34.0
8    40.0
9    30.0
Name: temperature, dtype: float64

In [20]:
df.temperature.mean()

32.666666666666664

In [16]:
df.temperature.fillna(df.temperature.mean())

0    32.000000
1    32.666667
2    28.000000
3    32.666667
4    32.000000
5    32.666667
6    32.666667
7    34.000000
8    40.000000
9    30.000000
Name: temperature, dtype: float64

In [49]:
df.temperature.fillna(df.temperature.mean(),inplace=True)

In [50]:
int(df.temperature.mean())

32

In [51]:
df

Date  temperature  windspeed  event
0 2022-01-01    32.000000        6.0   Rain
1 2022-01-02    32.666667        9.0  Sunny
2 2022-01-03    28.000000        8.5   Snow
3 2022-01-04    32.666667        7.0  Sunny
4 2022-01-05    32.000000        8.5   Rain
5 2022-01-06    32.666667        8.5  Sunny
6 2022-01-07    32.666667        8.5  Sunny
7 2022-01-08    34.000000        8.0  Cloud
8 2022-01-09    40.000000       12.0  Sunny
9 2022-01-10    30.000000       10.0  Sunny

In [23]:
df.windspeed.median()

8.5

In [24]:
df.windspeed.fillna(df.windspeed.median())

0     6.0
1     9.0
2     8.5
3     7.0
4     8.5
5     8.5
6     8.5
7     8.0
8    12.0
9    10.0
Name: windspeed, dtype: float64

In [26]:
df.windspeed.fillna(df.windspeed.median(),inplace=True)

In [27]:
df

Date  temperature  windspeed  event
0 2022-01-01    32.000000        6.0   Rain
1 2022-01-02    32.666667        9.0  Sunny
2 2022-01-03    28.000000        8.5   Snow
3 2022-01-04    32.666667        7.0    NaN
4 2022-01-05    32.000000        8.5   Rain
5 2022-01-06    32.666667        8.5  Sunny
6 2022-01-07    32.666667        8.5    NaN
7 2022-01-08    34.000000        8.0  Cloud
8 2022-01-09    40.000000       12.0  Sunny
9 2022-01-10    30.000000       10.0  Sunny

In [32]:
df.event.mode().values[0]

'Sunny'

In [33]:
df.event.fillna(df.event.mode().values[0])

0     Rain
1    Sunny
2     Snow
3    Sunny
4     Rain
5    Sunny
6    Sunny
7    Cloud
8    Sunny
9    Sunny
Name: event, dtype: object

In [35]:
df.event.fillna(df.event.mode().values[0],inplace=True)

In [36]:
df

Date  temperature  windspeed  event
0 2022-01-01    32.000000        6.0   Rain
1 2022-01-02    32.666667        9.0  Sunny
2 2022-01-03    28.000000        8.5   Snow
3 2022-01-04    32.666667        7.0  Sunny
4 2022-01-05    32.000000        8.5   Rain
5 2022-01-06    32.666667        8.5  Sunny
6 2022-01-07    32.666667        8.5  Sunny
7 2022-01-08    34.000000        8.0  Cloud
8 2022-01-09    40.000000       12.0  Sunny
9 2022-01-10    30.000000       10.0  Sunny

In [21]:
import numpy as np
np.median([6,7,8,9,10,12])

8.5

In [40]:
np.round(df.temperature.mean(),2)

32.67

In [41]:
int(df.temperature.mean())

32

In [52]:
df2=pd.read_excel("wheather.xlsx")

In [53]:
df2

Date  temperature  windspeed  event
0 2022-01-01         32.0        6.0   Rain
1 2022-01-02          NaN        9.0  Sunny
2 2022-01-03         28.0        NaN   Snow
3 2022-01-04          NaN        7.0    NaN
4 2022-01-05         32.0        NaN   Rain
5 2022-01-06          NaN        NaN  Sunny
6 2022-01-07          NaN        NaN    NaN
7 2022-01-08         34.0        8.0  Cloud
8 2022-01-09         40.0       12.0  Sunny
9 2022-01-10         30.0       10.0  Sunny

In [56]:
df2.temperature.ffill(inplace=True)

In [57]:
df2

Date  temperature  windspeed  event
0 2022-01-01         32.0        6.0   Rain
1 2022-01-02         32.0        9.0  Sunny
2 2022-01-03         28.0        NaN   Snow
3 2022-01-04         28.0        7.0    NaN
4 2022-01-05         32.0        NaN   Rain
5 2022-01-06         32.0        NaN  Sunny
6 2022-01-07         32.0        NaN    NaN
7 2022-01-08         34.0        8.0  Cloud
8 2022-01-09         40.0       12.0  Sunny
9 2022-01-10         30.0       10.0  Sunny

In [60]:
df2.windspeed.backfill(inplace=True)

C:\Users\Ducat\AppData\Local\Temp\ipykernel_4868\2528436913.py:1: FutureWarning: DataFrame.backfill/Series.backfill is deprecated. Use DataFrame.bfill/Series.bfill instead
  df2.windspeed.backfill(inplace=True)


In [61]:
df2

Date  temperature  windspeed  event
0 2022-01-01         32.0        6.0   Rain
1 2022-01-02         32.0        9.0  Sunny
2 2022-01-03         28.0        7.0   Snow
3 2022-01-04         28.0        7.0    NaN
4 2022-01-05         32.0        8.0   Rain
5 2022-01-06         32.0        8.0  Sunny
6 2022-01-07         32.0        8.0    NaN
7 2022-01-08         34.0        8.0  Cloud
8 2022-01-09         40.0       12.0  Sunny
9 2022-01-10         30.0       10.0  Sunny

In [65]:
df2.event.bfill(inplace=True)

In [66]:
df2

Date  temperature  windspeed  event
0 2022-01-01         32.0        6.0   Rain
1 2022-01-02         32.0        9.0  Sunny
2 2022-01-03         28.0        7.0   Snow
3 2022-01-04         28.0        7.0   Rain
4 2022-01-05         32.0        8.0   Rain
5 2022-01-06         32.0        8.0  Sunny
6 2022-01-07         32.0        8.0  Cloud
7 2022-01-08         34.0        8.0  Cloud
8 2022-01-09         40.0       12.0  Sunny
9 2022-01-10         30.0       10.0  Sunny

In [1]:
# import tips dataset

In [1]:
import pandas as pd
df3=pd.read_csv("tips (1).csv")

In [2]:
df3

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [3]:
df3.total_bill.sum()

4827.77

In [4]:
df3.tip.sum()

731.5799999999999

In [5]:
sum(df3.total_bill)

4827.770000000001

In [6]:
# to check is there any missing value of not


In [7]:
df3.isnull().sum()

total_bill    0
tip           0
sex           0
smoker        0
day           0
time          0
size          0
dtype: int64

In [8]:
df3

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [9]:
df3.day

0       Sun
1       Sun
2       Sun
3       Sun
4       Sun
       ... 
239     Sat
240     Sat
241     Sat
242     Sat
243    Thur
Name: day, Length: 244, dtype: object

In [10]:
df3.day.value_counts()

day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64

In [11]:
df3.day.value_counts().index

Index(['Sat', 'Sun', 'Thur', 'Fri'], dtype='object', name='day')

In [12]:
df3.day.value_counts().values

array([87, 76, 62, 19], dtype=int64)

In [13]:
df3.time.value_counts()

time
Dinner    176
Lunch      68
Name: count, dtype: int64

In [14]:
df3.smoker.value_counts()

smoker
No     151
Yes     93
Name: count, dtype: int64

In [15]:
df3.sex

0      Female
1        Male
2        Male
3        Male
4      Female
        ...  
239      Male
240    Female
241      Male
242      Male
243    Female
Name: sex, Length: 244, dtype: object

In [16]:
df3.sex.value_counts()

sex
Male      157
Female     87
Name: count, dtype: int64

In [17]:
df3.total_bill

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [18]:
import numpy as np
np.max(df3.total_bill)

50.81

In [19]:
np.min(df3.total_bill)

3.07

In [20]:
df3.tip

0      1.01
1      1.66
2      3.50
3      3.31
4      3.61
       ... 
239    5.92
240    2.00
241    2.00
242    1.75
243    3.00
Name: tip, Length: 244, dtype: float64

In [21]:
np.max(df3.tip)

10.0

In [22]:
np.min(df3.tip)

1.0

In [23]:
df3

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [24]:
np.sum(df3["size"])

627

In [25]:
# conditional

In [26]:
df3[df3.sex=="Male"].shape

(157, 7)

In [27]:
df3[(df3.sex=="Male")&(df3.smoker=="Yes")]

total_bill    tip   sex smoker   day    time  size
56        38.01   3.00  Male    Yes   Sat  Dinner     4
58        11.24   1.76  Male    Yes   Sat  Dinner     2
60        20.29   3.21  Male    Yes   Sat  Dinner     2
61        13.81   2.00  Male    Yes   Sat  Dinner     2
62        11.02   1.98  Male    Yes   Sat  Dinner     2
63        18.29   3.76  Male    Yes   Sat  Dinner     4
69        15.01   2.09  Male    Yes   Sat  Dinner     2
76        17.92   3.08  Male    Yes   Sat  Dinner     2
80        19.44   3.00  Male    Yes  Thur   Lunch     2
83        32.68   5.00  Male    Yes  Thur   Lunch     2
90        28.97   3.00  Male    Yes   Fri  Dinner     2
95        40.17   4.73  Male    Yes   Fri  Dinner     4
96        27.28   4.00  Male    Yes   Fri  Dinner     2
97        12.03   1.50  Male    Yes   Fri  Dinner     2
98        21.01   3.00  Male    Yes   Fri  Dinner     2
105       15.36   1.64  Male    Yes   Sat  Dinner     2
106       20.49   4.06  Male    Yes   Sat  Dinner     2
107       25.21   4.29  Male    Yes   Sat  Dinner     2
138       16.00   2.00  Male    Yes  Thur   Lunch     2
170       50.81  10.00  Male    Yes   Sat  Dinner     3
171       15.81   3.16  Male    Yes   Sat  Dinner     2
172        7.25   5.15  Male    Yes   Sun  Dinner     2
173       31.85   3.18  Male    Yes   Sun  Dinner     2
174       16.82   4.00  Male    Yes   Sun  Dinner     2
175       32.90   3.11  Male    Yes   Sun  Dinner     2
176       17.89   2.00  Male    Yes   Sun  Dinner     2
177       14.48   2.00  Male    Yes   Sun  Dinner     2
179       34.63   3.55  Male    Yes   Sun  Dinner     2
180       34.65   3.68  Male    Yes   Sun  Dinner     4
181       23.33   5.65  Male    Yes   Sun  Dinner     2
182       45.35   3.50  Male    Yes   Sun  Dinner     3
183       23.17   6.50  Male    Yes   Sun  Dinner     4
184       40.55   3.00  Male    Yes   Sun  Dinner     2
187       30.46   2.00  Male    Yes   Sun  Dinner     5
189       23.10   4.00  Male    Yes   Sun  Dinner     3
190       15.69   1.50  Male    Yes   Sun  Dinner     2
192       28.44   2.56  Male    Yes  Thur   Lunch     2
193       15.48   2.02  Male    Yes  Thur   Lunch     2
194       16.58   4.00  Male    Yes  Thur   Lunch     2
196       10.34   2.00  Male    Yes  Thur   Lunch     2
199       13.51   2.00  Male    Yes  Thur   Lunch     2
200       18.71   4.00  Male    Yes  Thur   Lunch     3
204       20.53   4.00  Male    Yes  Thur   Lunch     4
206       26.59   3.41  Male    Yes   Sat  Dinner     3
207       38.73   3.00  Male    Yes   Sat  Dinner     4
208       24.27   2.03  Male    Yes   Sat  Dinner     2
210       30.06   2.00  Male    Yes   Sat  Dinner     3
211       25.89   5.16  Male    Yes   Sat  Dinner     4
216       28.15   3.00  Male    Yes   Sat  Dinner     5
217       11.59   1.50  Male    Yes   Sat  Dinner     2
218        7.74   1.44  Male    Yes   Sat  Dinner     2
220       12.16   2.20  Male    Yes   Fri   Lunch     2
222        8.58   1.92  Male    Yes   Fri   Lunch     1
224       13.42   1.58  Male    Yes   Fri   Lunch     2
230       24.01   2.00  Male    Yes   Sat  Dinner     4
231       15.69   3.00  Male    Yes   Sat  Dinner     3
234       15.53   3.00  Male    Yes   Sat  Dinner     2
236       12.60   1.00  Male    Yes   Sat  Dinner     2
237       32.83   1.17  Male    Yes   Sat  Dinner     2
241       22.67   2.00  Male    Yes   Sat  Dinner     2

In [28]:
df3[(df3.sex=="Male")&(df3.smoker=="Yes")].shape  # & it is known as "and"

(60, 7)

In [29]:
(60/157)*100

38.21656050955414

In [30]:
df3[df3.sex=="Female"]

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
11        35.26  5.00  Female     No   Sun  Dinner     4
14        14.83  3.02  Female     No   Sun  Dinner     2
16        10.33  1.67  Female     No   Sun  Dinner     3
..          ...   ...     ...    ...   ...     ...   ...
226       10.09  2.00  Female    Yes   Fri   Lunch     2
229       22.12  2.88  Female    Yes   Sat  Dinner     2
238       35.83  4.67  Female     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[87 rows x 7 columns]

In [31]:
df3[df3.sex=="Female"].shape

(87, 7)

In [32]:
df3[(df3.sex=="Female")&(df3.smoker=="Yes")]

total_bill   tip     sex smoker   day    time  size
67         3.07  1.00  Female    Yes   Sat  Dinner     1
72        26.86  3.14  Female    Yes   Sat  Dinner     2
73        25.28  5.00  Female    Yes   Sat  Dinner     2
92         5.75  1.00  Female    Yes   Fri  Dinner     2
93        16.32  4.30  Female    Yes   Fri  Dinner     2
100       11.35  2.50  Female    Yes   Fri  Dinner     2
101       15.38  3.00  Female    Yes   Fri  Dinner     2
102       44.30  2.50  Female    Yes   Sat  Dinner     3
103       22.42  3.48  Female    Yes   Sat  Dinner     2
109       14.31  4.00  Female    Yes   Sat  Dinner     2
164       17.51  3.00  Female    Yes   Sun  Dinner     2
168       10.59  1.61  Female    Yes   Sat  Dinner     2
169       10.63  2.00  Female    Yes   Sat  Dinner     2
178        9.60  4.00  Female    Yes   Sun  Dinner     2
186       20.90  3.50  Female    Yes   Sun  Dinner     3
188       18.15  3.50  Female    Yes   Sun  Dinner     3
191       19.81  4.19  Female    Yes  Thur   Lunch     2
197       43.11  5.00  Female    Yes  Thur   Lunch     4
198       13.00  2.00  Female    Yes  Thur   Lunch     2
201       12.74  2.01  Female    Yes  Thur   Lunch     2
202       13.00  2.00  Female    Yes  Thur   Lunch     2
203       16.40  2.50  Female    Yes  Thur   Lunch     2
205       16.47  3.23  Female    Yes  Thur   Lunch     3
209       12.76  2.23  Female    Yes   Sat  Dinner     2
213       13.27  2.50  Female    Yes   Sat  Dinner     2
214       28.17  6.50  Female    Yes   Sat  Dinner     3
215       12.90  1.10  Female    Yes   Sat  Dinner     2
219       30.14  3.09  Female    Yes   Sat  Dinner     4
221       13.42  3.48  Female    Yes   Fri   Lunch     2
225       16.27  2.50  Female    Yes   Fri   Lunch     2
226       10.09  2.00  Female    Yes   Fri   Lunch     2
229       22.12  2.88  Female    Yes   Sat  Dinner     2
240       27.18  2.00  Female    Yes   Sat  Dinner     2

In [33]:
df3[(df3.sex=="Female")&(df3.smoker=="Yes")].shape

(33, 7)

In [34]:
(33/87)*100

37.93103448275862

In [35]:
# Report1: 
print("Total % of male smoker :",38.21656050955414)
print("Total % of female smoker :", 37.93103448275862)

Total % of male smoker : 38.21656050955414
Total % of female smoker : 37.93103448275862


In [36]:
df3

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [38]:
df3.day.value_counts()

day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64

In [39]:
df3[(df3.day=="Sat")&(df3.sex=="Female")]

total_bill   tip     sex smoker  day    time  size
21        20.29  2.75  Female     No  Sat  Dinner     2
22        15.77  2.23  Female     No  Sat  Dinner     2
29        19.65  3.00  Female     No  Sat  Dinner     2
32        15.06  3.00  Female     No  Sat  Dinner     2
33        20.69  2.45  Female     No  Sat  Dinner     4
37        16.93  3.07  Female     No  Sat  Dinner     3
57        26.41  1.50  Female     No  Sat  Dinner     2
66        16.45  2.47  Female     No  Sat  Dinner     2
67         3.07  1.00  Female    Yes  Sat  Dinner     1
71        17.07  3.00  Female     No  Sat  Dinner     3
72        26.86  3.14  Female    Yes  Sat  Dinner     2
73        25.28  5.00  Female    Yes  Sat  Dinner     2
74        14.73  2.20  Female     No  Sat  Dinner     2
102       44.30  2.50  Female    Yes  Sat  Dinner     3
103       22.42  3.48  Female    Yes  Sat  Dinner     2
104       20.92  4.08  Female     No  Sat  Dinner     2
109       14.31  4.00  Female    Yes  Sat  Dinner     2
111        7.25  1.00  Female     No  Sat  Dinner     1
168       10.59  1.61  Female    Yes  Sat  Dinner     2
169       10.63  2.00  Female    Yes  Sat  Dinner     2
209       12.76  2.23  Female    Yes  Sat  Dinner     2
213       13.27  2.50  Female    Yes  Sat  Dinner     2
214       28.17  6.50  Female    Yes  Sat  Dinner     3
215       12.90  1.10  Female    Yes  Sat  Dinner     2
219       30.14  3.09  Female    Yes  Sat  Dinner     4
229       22.12  2.88  Female    Yes  Sat  Dinner     2
238       35.83  4.67  Female     No  Sat  Dinner     3
240       27.18  2.00  Female    Yes  Sat  Dinner     2

In [40]:
df3[(df3.day=="Sat")&(df3.sex=="Female")].shape

(28, 7)

In [42]:
df4=df3[(df3.day=="Sat")&(df3.sex=="Female")]

In [43]:
df4

total_bill   tip     sex smoker  day    time  size
21        20.29  2.75  Female     No  Sat  Dinner     2
22        15.77  2.23  Female     No  Sat  Dinner     2
29        19.65  3.00  Female     No  Sat  Dinner     2
32        15.06  3.00  Female     No  Sat  Dinner     2
33        20.69  2.45  Female     No  Sat  Dinner     4
37        16.93  3.07  Female     No  Sat  Dinner     3
57        26.41  1.50  Female     No  Sat  Dinner     2
66        16.45  2.47  Female     No  Sat  Dinner     2
67         3.07  1.00  Female    Yes  Sat  Dinner     1
71        17.07  3.00  Female     No  Sat  Dinner     3
72        26.86  3.14  Female    Yes  Sat  Dinner     2
73        25.28  5.00  Female    Yes  Sat  Dinner     2
74        14.73  2.20  Female     No  Sat  Dinner     2
102       44.30  2.50  Female    Yes  Sat  Dinner     3
103       22.42  3.48  Female    Yes  Sat  Dinner     2
104       20.92  4.08  Female     No  Sat  Dinner     2
109       14.31  4.00  Female    Yes  Sat  Dinner     2
111        7.25  1.00  Female     No  Sat  Dinner     1
168       10.59  1.61  Female    Yes  Sat  Dinner     2
169       10.63  2.00  Female    Yes  Sat  Dinner     2
209       12.76  2.23  Female    Yes  Sat  Dinner     2
213       13.27  2.50  Female    Yes  Sat  Dinner     2
214       28.17  6.50  Female    Yes  Sat  Dinner     3
215       12.90  1.10  Female    Yes  Sat  Dinner     2
219       30.14  3.09  Female    Yes  Sat  Dinner     4
229       22.12  2.88  Female    Yes  Sat  Dinner     2
238       35.83  4.67  Female     No  Sat  Dinner     3
240       27.18  2.00  Female    Yes  Sat  Dinner     2

In [45]:
df4.total_bill.shape

(28,)

In [47]:
sum(df4.total_bill)

551.0499999999998

In [48]:
df3[(df3.sex=="Male")&(df3.day=="Sat")]

total_bill    tip   sex smoker  day    time  size
19        20.65   3.35  Male     No  Sat  Dinner     3
20        17.92   4.08  Male     No  Sat  Dinner     2
23        39.42   7.58  Male     No  Sat  Dinner     4
24        19.82   3.18  Male     No  Sat  Dinner     2
25        17.81   2.34  Male     No  Sat  Dinner     4
26        13.37   2.00  Male     No  Sat  Dinner     2
27        12.69   2.00  Male     No  Sat  Dinner     2
28        21.70   4.30  Male     No  Sat  Dinner     2
30         9.55   1.45  Male     No  Sat  Dinner     2
31        18.35   2.50  Male     No  Sat  Dinner     4
34        17.78   3.27  Male     No  Sat  Dinner     2
35        24.06   3.60  Male     No  Sat  Dinner     3
36        16.31   2.00  Male     No  Sat  Dinner     3
38        18.69   2.31  Male     No  Sat  Dinner     3
39        31.27   5.00  Male     No  Sat  Dinner     3
40        16.04   2.24  Male     No  Sat  Dinner     3
56        38.01   3.00  Male    Yes  Sat  Dinner     4
58        11.24   1.76  Male    Yes  Sat  Dinner     2
59        48.27   6.73  Male     No  Sat  Dinner     4
60        20.29   3.21  Male    Yes  Sat  Dinner     2
61        13.81   2.00  Male    Yes  Sat  Dinner     2
62        11.02   1.98  Male    Yes  Sat  Dinner     2
63        18.29   3.76  Male    Yes  Sat  Dinner     4
64        17.59   2.64  Male     No  Sat  Dinner     3
65        20.08   3.15  Male     No  Sat  Dinner     3
68        20.23   2.01  Male     No  Sat  Dinner     2
69        15.01   2.09  Male    Yes  Sat  Dinner     2
70        12.02   1.97  Male     No  Sat  Dinner     2
75        10.51   1.25  Male     No  Sat  Dinner     2
76        17.92   3.08  Male    Yes  Sat  Dinner     2
105       15.36   1.64  Male    Yes  Sat  Dinner     2
106       20.49   4.06  Male    Yes  Sat  Dinner     2
107       25.21   4.29  Male    Yes  Sat  Dinner     2
108       18.24   3.76  Male     No  Sat  Dinner     2
110       14.00   3.00  Male     No  Sat  Dinner     2
170       50.81  10.00  Male    Yes  Sat  Dinner     3
171       15.81   3.16  Male    Yes  Sat  Dinner     2
206       26.59   3.41  Male    Yes  Sat  Dinner     3
207       38.73   3.00  Male    Yes  Sat  Dinner     4
208       24.27   2.03  Male    Yes  Sat  Dinner     2
210       30.06   2.00  Male    Yes  Sat  Dinner     3
211       25.89   5.16  Male    Yes  Sat  Dinner     4
212       48.33   9.00  Male     No  Sat  Dinner     4
216       28.15   3.00  Male    Yes  Sat  Dinner     5
217       11.59   1.50  Male    Yes  Sat  Dinner     2
218        7.74   1.44  Male    Yes  Sat  Dinner     2
227       20.45   3.00  Male     No  Sat  Dinner     4
228       13.28   2.72  Male     No  Sat  Dinner     2
230       24.01   2.00  Male    Yes  Sat  Dinner     4
231       15.69   3.00  Male    Yes  Sat  Dinner     3
232       11.61   3.39  Male     No  Sat  Dinner     2
233       10.77   1.47  Male     No  Sat  Dinner     2
234       15.53   3.00  Male    Yes  Sat  Dinner     2
235       10.07   1.25  Male     No  Sat  Dinner     2
236       12.60   1.00  Male    Yes  Sat  Dinner     2
237       32.83   1.17  Male    Yes  Sat  Dinner     2
239       29.03   5.92  Male     No  Sat  Dinner     3
241       22.67   2.00  Male    Yes  Sat  Dinner     2
242       17.82   1.75  Male     No  Sat  Dinner     2

In [52]:
df5=df3[(df3.sex=="Male")&(df3.day=="Sat")]

In [54]:
df5.total_bill

19     20.65
20     17.92
23     39.42
24     19.82
25     17.81
26     13.37
27     12.69
28     21.70
30      9.55
31     18.35
34     17.78
35     24.06
36     16.31
38     18.69
39     31.27
40     16.04
56     38.01
58     11.24
59     48.27
60     20.29
61     13.81
62     11.02
63     18.29
64     17.59
65     20.08
68     20.23
69     15.01
70     12.02
75     10.51
76     17.92
105    15.36
106    20.49
107    25.21
108    18.24
110    14.00
170    50.81
171    15.81
206    26.59
207    38.73
208    24.27
210    30.06
211    25.89
212    48.33
216    28.15
217    11.59
218     7.74
227    20.45
228    13.28
230    24.01
231    15.69
232    11.61
233    10.77
234    15.53
235    10.07
236    12.60
237    32.83
239    29.03
241    22.67
242    17.82
Name: total_bill, dtype: float64

In [61]:
#mean of male payment on saturday
df5.total_bill.mean()

20.802542372881355

In [65]:
df4.total_bill.mean()

19.68035714285714

Report2: 

Average payment of man on saturday is 20.802542372881355

Average payment of female on saturday is 19.68035714285714

In [68]:
# motive: to find male and female average payment according to time.
df3

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [71]:
df3[(df3.time=="Lunch")&(df3.sex=="Male")&(df3.day=="Sat")]

Empty DataFrame
Columns: [total_bill, tip, sex, smoker, day, time, size]
Index: []

In [72]:
df3[(df3.time=="Lunch")&(df3.sex=="Female")&(df3.day=="Sat")]

Empty DataFrame
Columns: [total_bill, tip, sex, smoker, day, time, size]
Index: []

In [74]:
df3[(df3.time=="Dinner")&(df3.sex=="Female")&(df3.day=="Sat")].shape

(28, 7)

Report3: 

One Saturday Restra was not open on Lunch time.

In [76]:
df3[(df3.time=="Dinner")&(df3.sex=="Female")&(df3.day=="Sun")].shape

(18, 7)

In [79]:
df3[(df3.time=="Dinner")&(df3.sex=="Female")&(df3.day=="Sun")].total_bill.sum()/18

19.87222222222222

In [80]:
df3[(df3.time=="Dinner")&(df3.sex=="Female")&(df3.day=="Sun")].total_bill.mean()

19.87222222222222

In [81]:
df3[(df3.time=="Dinner")&(df3.sex=="Male")&(df3.day=="Sun")].total_bill.mean()

21.887241379310346

In [99]:
df3.total_bill.groupby(by=[df3.sex,df3.day,df3.time,df3.smoker,df3["size"]]).agg(["sum","min","max","mean","count"])

sum    min    max       mean  count
sex    day  time   smoker size                                        
Female Fri  Dinner No     2      22.75  22.75  22.75  22.750000      1
                   Yes    2      48.80   5.75  16.32  12.200000      4
            Lunch  No     3      15.98  15.98  15.98  15.980000      1
                   Yes    2      39.78  10.09  16.27  13.260000      3
       Sat  Dinner No     1       7.25   7.25   7.25   7.250000      1
                          2     149.28  14.73  26.41  18.660000      8
                          3      69.83  16.93  35.83  23.276667      3
                          4      20.69  20.69  20.69  20.690000      1
                   Yes    1       3.07   3.07   3.07   3.070000      1
                          2     198.32  10.59  27.18  18.029091     11
                          3      72.47  28.17  44.30  36.235000      2
                          4      30.14  30.14  30.14  30.140000      1
       Sun  Dinner No     2      72.81  10.29  17.31  14.562000      5
                          3      69.22  10.33  25.71  17.305000      4
                          4     119.66  24.59  35.26  29.915000      4
                          5      29.85  29.85  29.85  29.850000      1
                   Yes    2      27.11   9.60  17.51  13.555000      2
                          3      39.05  18.15  20.90  19.525000      2
       Thur Dinner No     2      18.78  18.78  18.78  18.780000      1
            Lunch  No     1      10.07  10.07  10.07  10.070000      1
                          2     237.11   8.35  20.27  13.172778     18
                          3      18.64  18.64  18.64  18.640000      1
                          4      58.91  24.08  34.83  29.455000      2
                          6      56.85  27.05  29.80  28.425000      2
                   Yes    2      74.95  12.74  19.81  14.990000      5
                          3      16.47  16.47  16.47  16.470000      1
                          4      43.11  43.11  43.11  43.110000      1
Male   Fri  Dinner No     2      34.95  12.46  22.49  17.475000      2
                   Yes    2      89.29  12.03  28.97  22.322500      4
                          4      40.17  40.17  40.17  40.170000      1
            Lunch  Yes    1       8.58   8.58   8.58   8.580000      1
                          2      25.58  12.16  13.42  12.790000      2
       Sat  Dinner No     2     251.38   9.55  21.70  14.787059     17
                          3     193.72  16.04  31.27  21.524444      9
                          4     192.63  17.81  48.33  32.105000      6
                   Yes    2     293.39   7.74  32.83  17.258235     17
                          3     123.15  15.69  50.81  30.787500      4
                          4     144.93  18.29  38.73  28.986000      5
                          5      28.15  28.15  28.15  28.150000      1
       Sun  Dinner No     2     349.53   8.77  23.95  15.887727     22
                          3     156.04  10.34  38.07  22.291429      7
                          4     302.91  16.49  32.40  25.242500     12
                          5      20.69  20.69  20.69  20.690000      1
                          6      48.17  48.17  48.17  48.170000      1
                   Yes    2     235.39   7.25  40.55  23.539000     10
                          3      68.45  23.10  45.35  34.225000      2
                          4      57.82  23.17  34.65  28.910000      2
                          5      30.46  30.46  30.46  30.460000      1
       Thur Lunch  No     2     244.22   7.51  24.71  15.263750     16
                          3      22.82  22.82  22.82  22.820000      1
                          4      27.20  27.20  27.20  27.200000      1
                          5      41.19  41.19  41.19  41.190000      1
                          6      34.30  34.30  34.30  34.300000      1
                   Yes    2     152.47  10.34  32.68  19.058750      8
                          3      18.71  18.71  18.71  18.71

In [100]:
df3.tip.groupby(by=[df3.sex,df3.day,df3.time,df3.smoker,df3["size"]]).agg(["sum","min","max","mean","count"])

sum   min    max      mean  count
sex    day  time   smoker size                                     
Female Fri  Dinner No     2      3.25  3.25   3.25  3.250000      1
                   Yes    2     10.80  1.00   4.30  2.700000      4
            Lunch  No     3      3.00  3.00   3.00  3.000000      1
                   Yes    2      7.98  2.00   3.48  2.660000      3
       Sat  Dinner No     1      1.00  1.00   1.00  1.000000      1
                          2     21.23  1.50   4.08  2.653750      8
                          3     10.74  3.00   4.67  3.580000      3
                          4      2.45  2.45   2.45  2.450000      1
                   Yes    1      1.00  1.00   1.00  1.000000      1
                          2     29.94  1.10   5.00  2.721818     11
                          3      9.00  2.50   6.50  4.500000      2
                          4      3.09  3.09   3.09  3.090000      1
       Sun  Dinner No     2     12.74  1.01   3.50  2.548000      5
                          3     11.17  1.67   4.00  2.792500      4
                          4     17.56  3.61   5.20  4.390000      4
                          5      5.14  5.14   5.14  5.140000      1
                   Yes    2      7.00  3.00   4.00  3.500000      2
                          3      7.00  3.50   3.50  3.500000      2
       Thur Dinner No     2      3.00  3.00   3.00  3.000000      1
            Lunch  No     1      1.83  1.83   1.83  1.830000      1
                          2     38.01  1.25   3.50  2.111667     18
                          3      1.36  1.36   1.36  1.360000      1
                          4      8.09  2.92   5.17  4.045000      2
                          6      9.20  4.20   5.00  4.600000      2
                   Yes    2     12.70  2.00   4.19  2.540000      5
                          3      3.23  3.23   3.23  3.230000      1
                          4      5.00  5.00   5.00  5.000000      1
Male   Fri  Dinner No     2      5.00  1.50   3.50  2.500000      2
                   Yes    2     11.50  1.50   4.00  2.875000      4
                          4      4.73  4.73   4.73  4.730000      1
            Lunch  Yes    1      1.92  1.92   1.92  1.920000      1
                          2      3.78  1.58   2.20  1.890000      2
       Sat  Dinner No     2     42.85  1.25   4.30  2.520588     17
                          3     30.21  2.00   5.92  3.356667      9
                          4     31.15  2.34   9.00  5.191667      6
                   Yes    2     39.41  1.00   4.29  2.318235     17
                          3     18.41  2.00  10.00  4.602500      4
                          4     16.92  2.00   5.16  3.384000      5
                          5      3.00  3.00   3.00  3.000000      1
       Sun  Dinner No     2     56.98  1.32   5.00  2.590000     22
                          3     21.14  1.66   4.00  3.020000      7
                          4     45.84  2.00   6.00  3.820000     12
                          5      5.00  5.00   5.00  5.000000      1
                          6      5.00  5.00   5.00  5.000000      1
                   Yes    2     33.14  1.50   5.65  3.314000     10
                          3      7.50  3.50   4.00  3.750000      2
                          4     10.18  3.68   6.50  5.090000      2
                          5      2.00  2.00   2.00  2.000000      1
       Thur Lunch  No     2     40.95  1.44   5.85  2.559375     16
                          3      2.18  2.18   2.18  2.180000      1
                          4      4.00  4.00   4.00  4.000000      1
                          5      5.00  5.00   5.00  5.000000      1
                          6      6.70  6.70   6.70  6.700000      1
                   Yes    2     22.58  2.00   5.00  2.822500      8
                          3      4.00  4.00   4.00  4.000000      1
                          4      4.00  4.00   4.00  4.000000      1

In [34]:
import pandas as pd

In [35]:
df1=pd.read_csv("Titanic.csv")

In [36]:
df1

PassengerId  Survived  Pclass   
0              1         0       3  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp   
0                              Braund, Mr. Owen Harris    male  22.0      1  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [37]:
df1.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [38]:
int(df1.Age.mean())

29

In [39]:
import numpy as np
np.round(df1.Age.mean())

30.0

In [40]:
df1.Age.fillna(np.round(df1.Age.mean()),inplace=True)

In [41]:
df1.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [42]:
df1

PassengerId  Survived  Pclass   
0              1         0       3  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp   
0                              Braund, Mr. Owen Harris    male  22.0      1  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  30.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [43]:
df1.iloc[0:891:1,[1,2,4,5,6,7,9,11]]

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0           0       3    male  22.0      1      0   7.2500        S
1           1       1  female  38.0      1      0  71.2833        C
2           1       3  female  26.0      0      0   7.9250        S
3           1       1  female  35.0      1      0  53.1000        S
4           0       3    male  35.0      0      0   8.0500        S
..        ...     ...     ...   ...    ...    ...      ...      ...
886         0       2    male  27.0      0      0  13.0000        S
887         1       1  female  19.0      0      0  30.0000        S
888         0       3  female  30.0      1      2  23.4500        S
889         1       1    male  26.0      0      0  30.0000        C
890         0       3    male  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

In [44]:
df2=df1.iloc[:,[1,2,4,5,6,7,9,11]]

In [45]:
df2

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0           0       3    male  22.0      1      0   7.2500        S
1           1       1  female  38.0      1      0  71.2833        C
2           1       3  female  26.0      0      0   7.9250        S
3           1       1  female  35.0      1      0  53.1000        S
4           0       3    male  35.0      0      0   8.0500        S
..        ...     ...     ...   ...    ...    ...      ...      ...
886         0       2    male  27.0      0      0  13.0000        S
887         1       1  female  19.0      0      0  30.0000        S
888         0       3  female  30.0      1      2  23.4500        S
889         1       1    male  26.0      0      0  30.0000        C
890         0       3    male  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

In [46]:
df2.Sex.value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [47]:
df2.Sex=df2.Sex.map({"male":0,"female":1})   #male assume as 0 and female assume as 1

C:\Users\Ducat\AppData\Local\Temp\ipykernel_3224\3283998145.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2.Sex=df2.Sex.map({"male":0,"female":1})   #male assume as 0 and female assume as 1


In [48]:
df2

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0           0       3    0  22.0      1      0   7.2500        S
1           1       1    1  38.0      1      0  71.2833        C
2           1       3    1  26.0      0      0   7.9250        S
3           1       1    1  35.0      1      0  53.1000        S
4           0       3    0  35.0      0      0   8.0500        S
..        ...     ...  ...   ...    ...    ...      ...      ...
886         0       2    0  27.0      0      0  13.0000        S
887         1       1    1  19.0      0      0  30.0000        S
888         0       3    1  30.0      1      2  23.4500        S
889         1       1    0  26.0      0      0  30.0000        C
890         0       3    0  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

In [49]:
df2.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    2
dtype: int64

In [50]:
df2.Embarked.value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

In [67]:
d=df2.Embarked.mode()
d

0    S
Name: Embarked, dtype: object

In [70]:
e=list(d)
e

['S']

In [71]:
e[0]

'S'

In [74]:
df2.Embarked=df2.Embarked.fillna(e[0])

C:\Users\Ducat\AppData\Local\Temp\ipykernel_3224\365082761.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2.Embarked=df2.Embarked.fillna(e[0])


In [75]:
df2.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64

In [76]:
df2

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0           0       3    0  22.0      1      0   7.2500        S
1           1       1    1  38.0      1      0  71.2833        C
2           1       3    1  26.0      0      0   7.9250        S
3           1       1    1  35.0      1      0  53.1000        S
4           0       3    0  35.0      0      0   8.0500        S
..        ...     ...  ...   ...    ...    ...      ...      ...
886         0       2    0  27.0      0      0  13.0000        S
887         1       1    1  19.0      0      0  30.0000        S
888         0       3    1  30.0      1      2  23.4500        S
889         1       1    0  26.0      0      0  30.0000        C
890         0       3    0  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

In [77]:
# Correlation

In [82]:
d=df2.iloc[:,:-1]
d

Survived  Pclass  Sex   Age  SibSp  Parch     Fare
0           0       3    0  22.0      1      0   7.2500
1           1       1    1  38.0      1      0  71.2833
2           1       3    1  26.0      0      0   7.9250
3           1       1    1  35.0      1      0  53.1000
4           0       3    0  35.0      0      0   8.0500
..        ...     ...  ...   ...    ...    ...      ...
886         0       2    0  27.0      0      0  13.0000
887         1       1    1  19.0      0      0  30.0000
888         0       3    1  30.0      1      2  23.4500
889         1       1    0  26.0      0      0  30.0000
890         0       3    0  32.0      0      0   7.7500

[891 rows x 7 columns]

In [83]:
d.corr()

Survived    Pclass       Sex       Age     SibSp     Parch      Fare
Survived  1.000000 -0.338481  0.543351 -0.070657 -0.035322  0.081629  0.257307
Pclass   -0.338481  1.000000 -0.131900 -0.329727  0.083081  0.018443 -0.549500
Sex       0.543351 -0.131900  1.000000 -0.084660  0.114631  0.245489  0.182333
Age      -0.070657 -0.329727 -0.084660  1.000000 -0.232440 -0.180330  0.090632
SibSp    -0.035322  0.083081  0.114631 -0.232440  1.000000  0.414838  0.159651
Parch     0.081629  0.018443  0.245489 -0.180330  0.414838  1.000000  0.216225
Fare      0.257307 -0.549500  0.182333  0.090632  0.159651  0.216225  1.000000